addon_service/credentials/encryption.py

```diff
@@ -0,0 +1,142 @@
+"""for encrypting and decrypting data (using secrets from settings)
+
+considering recommendations from:
+- https://cryptography.io/en/latest/fernet/
+- https://docs.python.org/3.12/library/hashlib.html#hashlib.scrypt
+- https://cryptography.io/en/latest/hazmat/primitives/key-derivation-functions/#scrypt
+- https://datatracker.ietf.org/doc/html/rfc7914
+- https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-132.pdf
+"""
+
+import base64
+import dataclasses
+import functools
+import hashlib
+import json
+import math
+import os
+
+from cryptography import fernet
+from django.conf import settings
+
+
+__all__ = (
+    "pls_decrypt_bytes",
+    "pls_decrypt_json",
+    "pls_encrypt_bytes",
+    "pls_encrypt_json",
+    "salt_factory",
+)
+
+
+# 32-byte key expected by cryptography.fernet.Fernet
+_KEY_BYTE_COUNT = 32
+
+# recommended len(salt) >= 16 bytes
+_SALT_BYTE_COUNT = settings.GRAVYVALET_SALT_BYTE_COUNT or 17
+
+
+def salt_factory() -> bytes:
+    return os.urandom(_SALT_BYTE_COUNT)
+
+
+@dataclasses.dataclass(frozen=True)  # frozen for use as lru_cache key
+class KeyParameters:  # https://datatracker.ietf.org/doc/html/rfc7914#section-2
+    salt: bytes = dataclasses.field(default_factory=salt_factory)
+    # recommended scrypt_cost ("N") between 2^14 and 2^20
+    scrypt_cost: int = settings.GRAVYVALET_SCRYPT_COST or 2**17
+    # recommended scrypt_block_size ("r") = 8
+    scrypt_block_size: int = settings.GRAVYVALET_SCRYPT_BLOCK_SIZE or 8
+    # recommended scrypt_parallelization ("p") = 1
+    scrypt_parallelization: int = settings.GRAVYVALET_SCRYPT_PARALLELIZATION or 3
```

Comment:
RFC 7914 recommends P=1. So should this be
```suggestion
    scrypt_parallelization: int = settings.GRAVYVALET_SCRYPT_PARALLELIZATION or 1
```

Reply:
hmm yeah, probably -- i bumped it arbitrarily after noticing django had [increased default parallelism to 5](https://code.djangoproject.com/ticket/35428), but on a closer look [the OWASP cheatsheet they cite](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#scrypt) recommends p=1 with our current default N=2^17 (tho it doesn't try to justify that recommendation 🤷 ) -- so yes, changed to 1